base/initdefs.jl

```diff
@@ -49,31 +49,84 @@ Determine whether Julia is running an interactive session.
 """
 isinteractive() = (is_interactive::Bool)
 
+## package depots (registries, packages, environments) ##
+
+const DEPOT_PATH = String[]
+
+function init_depot_path(BINDIR = Sys.BINDIR)
+    if haskey(ENV, "JULIA_DEPOT_PATH")
+        depots = split(ENV["JULIA_DEPOT_PATH"], Sys.iswindows() ? ';' : ':')
+        push!(empty!(DEPOT_PATH), map(expanduser, depots))
+    else
+        push!(DEPOT_PATH, joinpath(homedir(), ".julia"))
+    end
+end
+
+## load-path types ##
+
+abstract type AbstractEnv end
+
+struct CurrentEnv <: AbstractEnv
+    create::Bool
+    CurrentEnv(; create::Bool=false) = new(create)
+end
+
+struct NamedEnv <: AbstractEnv
+    name::String
+    create::Bool
+    NamedEnv(name::String; create::Bool=false) = new(name, create)
+end
+
+function parse_env(env::Union{String,SubString{String}})
+    env == "@" && return CurrentEnv()
+    env == "@!" && return CurrentEnv(create=true)
+    if env[1] == '@'
+        create = env[2] == '!'
+        name = env[2+create:end]
+        name = replace(name, '#' => VERSION.major, count=1)
+        name = replace(name, '#' => VERSION.minor, count=1)
+        name = replace(name, '#' => VERSION.patch, count=1)
+        return NamedEnv(name, create=create)
+    end
+    return env # literal path
+end
+
 """
     LOAD_PATH
 
 An array of paths as strings or custom loader objects for the `require`
 function and `using` and `import` statements to consider when loading
 code.
 """
-const LOAD_PATH = String[]
+const LOAD_PATH = Any[]
 const LOAD_CACHE_PATH = String[]
 
+function parse_load_path(str::String)
+    envs = Any[split(str, Sys.iswindows() ? ';' : ':');]
+    for (i, env) in enumerate(envs)
+        if '|' in env
+            envs[i] = [parse_env(e) for e in split(env, '|')]
+        else
+            envs[i] = parse_env(env)
+        end
+    end
+    return envs
+end
+
 function init_load_path(BINDIR = Sys.BINDIR)
+    load_path = get(ENV, "JULIA_LOAD_PATH", "@|@v#.#.#|@v#.#|@v#|@default|@!v#.#")
```

Comment:
For other readers see https://github.com/JuliaLang/Pkg3.jl/issues/82 for context.
The string `"@|@v#.#.#|@v#.#|@v#|@default|@!v#.#"` turns into:

```
LOAD_PATH = Any[
    [ CurrentProject(),
      NamedEnv(“v$(VERSION.major).$(VERSION.minor).$(VERSION.patch)”),
      NamedEnv(“v$(VERSION.major).$(VERSION.minor)”),
      NamedEnv(“v$(VERSION.major)”),
      NamedEnv(“default”),
      NamedEnv(“v$(VERSION.major).$(VERSION.minor)”, create=true),
]
```

Reply:
Yes, we need to bikeshed this syntax, but this is what I ended up going with:

* `@` = current project environment found by looking in parent directories
* `@name` = named environment if it already exists
* `@!name` = named environment even if it doesn't already exist
* `#` characters are replaced by the major, minor and patch version numbers in order

I decided not to introduce any escaping since you can always use `.juliarc.jl` to set the values programmatically. If you have an actual relative path name that starts with a `@` you can always write it as `./@...`